# Dealing with Categorical Data

Many machine learning models can not deal with categorical data set as strings. For example linear regression can not apply a a Beta Coefficent to colors like "red" or "blue". Instead we need to convert these categories into "dummy" variables, otherwise known as "one-hot" encoding.

## Imports

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import warnings

In [3]:
warnings.simplefilter(action='ignore', category=FutureWarning)

## Data

We will open a .csv file that has been "cleaned" to remove outliers and NaN

In [4]:
df = pd.read_csv('./data/Ames_NO_Missing_Data.csv')

In [10]:
pd.set_option('display.max_columns', None)

In [11]:
df.head()

MS SubClass MS Zoning  Lot Frontage  Lot Area Street Lot Shape Land Contour  \
0          20        RL         141.0     31770   Pave       IR1          Lvl   
1          20        RH          80.0     11622   Pave       Reg          Lvl   
2          20        RL          81.0     14267   Pave       IR1          Lvl   
3          20        RL          93.0     11160   Pave       Reg          Lvl   
4          60        RL          74.0     13830   Pave       IR1          Lvl   

  Utilities Lot Config Land Slope Neighborhood Condition 1 Condition 2  \
0    AllPub     Corner        Gtl        NAmes        Norm        Norm   
1    AllPub     Inside        Gtl        NAmes       Feedr        Norm   
2    AllPub     Corner        Gtl        NAmes        Norm        Norm   
3    AllPub     Corner        Gtl        NAmes        Norm        Norm   
4    AllPub     Inside        Gtl      Gilbert        Norm        Norm   

  Bldg Type House Style  Overall Qual  Overall Cond  Year Built  \
0      1Fam      1Story             6             5        1960   
1      1Fam      1Story             5             6        1961   
2      1Fam      1Story             6             6        1958   
3      1Fam      1Story             7             5        1968   
4      1Fam      2Story             5             5        1997   

   Year Remod/Add Roof Style Roof Matl Exterior 1st Exterior 2nd Mas Vnr Type  \
0            1960        Hip   CompShg      BrkFace      Plywood        Stone   
1            1961      Gable   CompShg      VinylSd      VinylSd          NaN   
2            1958        Hip   CompShg      Wd Sdng      Wd Sdng      BrkFace   
3            1968        Hip   CompShg      BrkFace      BrkFace          NaN   
4            1998      Gable   CompShg      VinylSd      VinylSd          NaN   

   Mas Vnr Area Exter Qual Exter Cond Foundation Bsmt Qual Bsmt Cond  \
0         112.0         TA         TA     CBlock        TA        Gd   
1           0.0         TA         TA     CBlock        TA        TA   
2         108.0         TA         TA     CBlock        TA        TA   
3           0.0         Gd         TA     CBlock        TA        TA   
4           0.0         TA         TA      PConc        Gd        TA   

  Bsmt Exposure BsmtFin Type 1  BsmtFin SF 1 BsmtFin Type 2  BsmtFin SF 2  \
0            Gd            BLQ         639.0            Unf           0.0   
1            No            Rec         468.0            LwQ         144.0   
2            No            ALQ         923.0            Unf           0.0   
3            No            ALQ        1065.0            Unf           0.0   
4            No            GLQ         791.0            Unf           0.0   

   Bsmt Unf SF  Total Bsmt SF Heating Heating QC Central Air Electrical  \
0        441.0         1080.0    GasA         Fa           Y      SBrkr   
1        270.0          882.0    GasA         TA           Y      SBrkr   
2        406.0         1329.0    GasA         TA           Y      SBrkr   
3       1045.0         2110.0    GasA         Ex           Y      SBrkr   
4        137.0          928.0    GasA         Gd           Y      SBrkr   

   1st Flr SF  2nd Flr SF  Low Qual Fin SF  Gr Liv Area  Bsmt Full Bath  \
0        1656           0                0         1656             1.0   
1         896           0                0          896             0.0   
2        1329           0                0         1329             0.0   
3        2110           0                0         2110             1.0   
4         928         701                0         1629             0.0   

   Bsmt Half Bath  Full Bath  Half Bath  Bedroom AbvGr  Kitchen AbvGr  \
0             0.0          1          0              3              1   
1             0.0          1          0              2              1   
2             0.0          1          1              3              1   
3             0.0          2          1              3              1   
4             0.0

# Data Description

In [6]:
with open('./data/Ames_Housing_Feature_Description.txt') as f:
    print(f.read())

MSSubClass: Identifies the type of dwelling involved in the sale.	

        20	1-STORY 1946 & NEWER ALL STYLES
        30	1-STORY 1945 & OLDER
        40	1-STORY W/FINISHED ATTIC ALL AGES
        45	1-1/2 STORY - UNFINISHED ALL AGES
        50	1-1/2 STORY FINISHED ALL AGES
        60	2-STORY 1946 & NEWER
        70	2-STORY 1945 & OLDER
        75	2-1/2 STORY ALL AGES
        80	SPLIT OR MULTI-LEVEL
        85	SPLIT FOYER
        90	DUPLEX - ALL STYLES AND AGES
       120	1-STORY PUD (Planned Unit Development) - 1946 & NEWER
       150	1-1/2 STORY PUD - ALL AGES
       160	2-STORY PUD - 1946 & NEWER
       180	PUD - MULTILEVEL - INCL SPLIT LEV/FOYER
       190	2 FAMILY CONVERSION - ALL STYLES AND AGES

MSZoning: Identifies the general zoning classification of the sale.
		
       A	Agriculture
       C	Commercial
       FV	Floating Village Residential
       I	Industrial
       RH	Residential High Density
       RL	Residential Low Density
       RP	Residential Low Density Park 
       RM

In [7]:
df.isnull().sum()

MS SubClass       0
MS Zoning         0
Lot Frontage      0
Lot Area          0
Street            0
                 ..
Mo Sold           0
Yr Sold           0
Sale Type         0
Sale Condition    0
SalePrice         0
Length: 76, dtype: int64

### Numerical Column to Categorical

## Numerical Column to Categorical

We need to be careful when it comes to encoding categories as numbers. We want to make sure that the numerical relationship makes sense for a model. For example, the encoding MSSubClass is essentially just a number code per class:

    MSSubClass: Identifies the type of dwelling involved in the sale.	

        20	1-STORY 1946 & NEWER ALL STYLES
        30	1-STORY 1945 & OLDER
        40	1-STORY W/FINISHED ATTIC ALL AGES
        45	1-1/2 STORY - UNFINISHED ALL AGES
        50	1-1/2 STORY FINISHED ALL AGES
        60	2-STORY 1946 & NEWER
        70	2-STORY 1945 & OLDER
        75	2-1/2 STORY ALL AGES
        80	SPLIT OR MULTI-LEVEL
        85	SPLIT FOYER
        90	DUPLEX - ALL STYLES AND AGES
       120	1-STORY PUD (Planned Unit Development) - 1946 & NEWER
       150	1-1/2 STORY PUD - ALL AGES
       160	2-STORY PUD - 1946 & NEWER
       180	PUD - MULTILEVEL - INCL SPLIT LEV/FOYER
       190	2 FAMILY CONVERSION - ALL STYLES AND AGES

The number itself does not appear to have a relationship to the other numbers. While 30 > 20 is True, it doesn't really make sense that "1-STORY 1945 & OLDER" > "1-STORY 1946 & NEWER ALL STYLES". Keep in mind, this isn't always the case, for example 1st class seats versus 2nd class seats encoded as 1 and 2. Make sure you fully understand your data set to examine what needs to be converted/changed.

### MSSubClass

In [34]:
# Convert to String
df['MS SubClass'] = df['MS SubClass'].apply(str)

## Creating "Dummy" Variables

### Avoiding MultiCollinearity and the Dummy Variable Trap

https://stats.stackexchange.com/questions/144372/dummy-variable-trap

### Creating Dummy Variables from Object Columns

In [12]:
df.select_dtypes(include='object')

MS SubClass MS Zoning Street Lot Shape Land Contour Utilities Lot Config  \
0             20        RL   Pave       IR1          Lvl    AllPub     Corner   
1             20        RH   Pave       Reg          Lvl    AllPub     Inside   
2             20        RL   Pave       IR1          Lvl    AllPub     Corner   
3             20        RL   Pave       Reg          Lvl    AllPub     Corner   
4             60        RL   Pave       IR1          Lvl    AllPub     Inside   
...          ...       ...    ...       ...          ...       ...        ...   
2920          80        RL   Pave       IR1          Lvl    AllPub    CulDSac   
2921          20        RL   Pave       IR1          Low    AllPub     Inside   
2922          85        RL   Pave       Reg          Lvl    AllPub     Inside   
2923          20        RL   Pave       Reg          Lvl    AllPub     Inside   
2924          60        RL   Pave       Reg          Lvl    AllPub     Inside   

     Land Slope Neighborhood Condition 1 Condition 2 Bldg Type House Style  \
0           Gtl        NAmes        Norm        Norm      1Fam      1Story   
1           Gtl        NAmes       Feedr        Norm      1Fam      1Story   
2           Gtl        NAmes        Norm        Norm      1Fam      1Story   
3           Gtl        NAmes        Norm        Norm      1Fam      1Story   
4           Gtl      Gilbert        Norm        Norm      1Fam      2Story   
...         ...          ...         ...         ...       ...         ...   
2920        Gtl      Mitchel        Norm        Norm      1Fam        SLvl   
2921        Mod      Mitchel        Norm        Norm      1Fam      1Story   
2922        Gtl      Mitchel        Norm        Norm      1Fam      SFoyer   
2923        Mod      Mitchel        Norm        Norm      1Fam      1Story   
2924        Mod      Mitchel        Norm        Norm      1Fam      2Story   

     Roof Style Roof Matl Exterior 1st Exterior 2nd Mas Vnr Type Exter Qual  \
0           Hip   CompShg      BrkFace      Plywood        Stone         TA   
1         Gable   CompShg      VinylSd      VinylSd          NaN         TA   
2           Hip   CompShg      Wd Sdng      Wd Sdng      BrkFace         TA   
3           Hip   CompShg      BrkFace      BrkFace          NaN         Gd   
4         Gable   CompShg      VinylSd      VinylSd          NaN         TA   
...         ...       ...          ...          ...          ...        ...   
2920      Gable   CompShg      HdBoard      HdBoard          NaN         TA   
2921      Gable   CompShg      HdBoard      HdBoard          NaN         TA   
2922      Gable   CompShg      HdBoard      Wd Shng          NaN         TA   
2923      Gable   CompShg      HdBoard      HdBoard          NaN         TA   
2924      Gable   CompShg      HdBoard      HdBoard      BrkFace         TA   

     Exter Cond Foundation Bsmt Qual Bsmt Cond Bsmt Exposure BsmtFin Type 1  \
0            TA     CBlock        TA        Gd            Gd            BLQ   
1            TA     CBlock        TA        TA            No            Rec   
2            TA     CBlock        TA        TA            No            ALQ   
3            TA     CBlock        TA        TA            No            ALQ   
4            TA      PConc        Gd        TA            No            GLQ   
...         ...        ...       ...       ...           ...            ...   
2920         TA     CBlock        TA        TA            Av            GLQ   
2921         TA     CBlock        Gd        TA            Av            BLQ   
2922         TA      PConc        Gd        TA            Av            GLQ   
2923         TA     CBlock        Gd        TA            Av            ALQ   
2924         TA      PConc        Gd        TA            Av            LwQ   

     BsmtFin Type 2 Heating Heating QC Central Air Electrical Kitchen Qual  \
0               Unf    GasA         Fa           Y      SBrkr           TA   
1               LwQ    GasA         TA   

In [35]:
df_objs = df.select_dtypes(include='object')
df_nums = df.select_dtypes(exclude='object')

In [36]:
df_nums.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2925 entries, 0 to 2924
Data columns (total 36 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Lot Frontage     2925 non-null   float64
 1   Lot Area         2925 non-null   int64  
 2   Overall Qual     2925 non-null   int64  
 3   Overall Cond     2925 non-null   int64  
 4   Year Built       2925 non-null   int64  
 5   Year Remod/Add   2925 non-null   int64  
 6   Mas Vnr Area     2925 non-null   float64
 7   BsmtFin SF 1     2925 non-null   float64
 8   BsmtFin SF 2     2925 non-null   float64
 9   Bsmt Unf SF      2925 non-null   float64
 10  Total Bsmt SF    2925 non-null   float64
 11  1st Flr SF       2925 non-null   int64  
 12  2nd Flr SF       2925 non-null   int64  
 13  Low Qual Fin SF  2925 non-null   int64  
 14  Gr Liv Area      2925 non-null   int64  
 15  Bsmt Full Bath   2925 non-null   float64
 16  Bsmt Half Bath   2925 non-null   float64
 17  Full Bath     

In [37]:
df_objs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2925 entries, 0 to 2924
Data columns (total 40 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   MS SubClass     2925 non-null   object
 1   MS Zoning       2925 non-null   object
 2   Street          2925 non-null   object
 3   Lot Shape       2925 non-null   object
 4   Land Contour    2925 non-null   object
 5   Utilities       2925 non-null   object
 6   Lot Config      2925 non-null   object
 7   Land Slope      2925 non-null   object
 8   Neighborhood    2925 non-null   object
 9   Condition 1     2925 non-null   object
 10  Condition 2     2925 non-null   object
 11  Bldg Type       2925 non-null   object
 12  House Style     2925 non-null   object
 13  Roof Style      2925 non-null   object
 14  Roof Matl       2925 non-null   object
 15  Exterior 1st    2925 non-null   object
 16  Exterior 2nd    2925 non-null   object
 17  Mas Vnr Type    1152 non-null   object
 18  Exter Qu

### Converting

In [39]:
df_objs = pd.get_dummies(df_objs, drop_first=True,dtype=float)

In [40]:
final_df = pd.concat([df_nums, df_objs], axis=1)

In [41]:
final_df

Lot Frontage  Lot Area  Overall Qual  Overall Cond  Year Built  \
0       141.000000     31770             6             5        1960   
1        80.000000     11622             5             6        1961   
2        81.000000     14267             6             6        1958   
3        93.000000     11160             7             5        1968   
4        74.000000     13830             5             5        1997   
...            ...       ...           ...           ...         ...   
2920     37.000000      7937             6             6        1984   
2921     75.144444      8885             5             5        1983   
2922     62.000000     10441             5             5        1992   
2923     77.000000     10010             5             5        1974   
2924     74.000000      9627             7             5        1993   

      Year Remod/Add  Mas Vnr Area  BsmtFin SF 1  BsmtFin SF 2  Bsmt Unf SF  \
0               1960         112.0         639.0           0.0        441.0   
1               1961           0.0         468.0         144.0        270.0   
2               1958         108.0         923.0           0.0        406.0   
3               1968           0.0        1065.0           0.0       1045.0   
4               1998           0.0         791.0           0.0        137.0   
...              ...           ...           ...           ...          ...   
2920            1984           0.0         819.0           0.0        184.0   
2921            1983           0.0         301.0         324.0        239.0   
2922            1992           0.0         337.0           0.0        575.0   
2923            1975           0.0        1071.0         123.0        195.0   
2924            1994          94.0         758.0           0.0        238.0   

      Total Bsmt SF  1st Flr SF  2nd Flr SF  Low Qual Fin SF  Gr Liv Area  \
0            1080.0        1656           0                0         1656   
1             882.0         896           0                0          896   
2            1329.0        1329           0                0         1329   
3            2110.0        2110           0                0         2110   
4             928.0         928         701                0         1629   
...             ...         ...         ...              ...          ...   
2920         1003.0        1003           0                0         1003   
2921          864.0         902           0                0          902   
2922          912.0         970           0                0          970   
2923         1389.0        1389           0                0         1389   
2924          996.0         996        1004                0         2000   

      Bsmt Full Bath  Bsmt Half Bath  Full Bath  Half Bath  Bedroom AbvGr  \
0                1.0             0.0          1          0              3   
1                0.0             0.0          1          0              2   
2                0.0             0.0          1          1              3   
3                1.0             0.0          2          1              3   
4                0.0             0.0          2          1              3   
...              ...             ...        ...        ...            ...   
2920             1.0             0.0          1          0              3   
2921             1.0             0.0          1          0              2   
2922             0.0             1.0          1          0              3   
2923             1.0             0.0          1          0              2   
2924             0.0             0.0          2          1              3   

      Kitchen AbvGr  TotRms AbvGrd  Fireplaces  Garage Yr Blt  Garage Cars  \
0                 1              7           2         1960.0          2.0   
1                 1              5           0         1961.0          1.0   
2                 1              6           0         1958.0          1.0   
3                 1           

## Final Thoughts

Keep in mind, we don't know if 263 columns are useful. More columns doesn't necessarily lead to better results. In fact, we may want to further remove columns (or later on use a model with regularization to choose important columns for us). What we have done here has greatly expanded the ratio of rows to columns, which may actually lead to worse performance (however you don't know until you've actually compared multiple models/approaches).

In [42]:
final_df.corr()['SalePrice'].sort_values()

Exter Qual_TA        -0.591459
Kitchen Qual_TA      -0.527461
Bsmt Qual_TA         -0.453022
Garage Finish_Unf    -0.422363
Garage Type_Detchd   -0.365209
                        ...   
Garage Cars           0.648488
Total Bsmt SF         0.660983
Gr Liv Area           0.727279
Overall Qual          0.802637
SalePrice             1.000000
Name: SalePrice, Length: 263, dtype: float64

    OverallQual: Rates the overall material and finish of the house

           10	Very Excellent
           9	Excellent
           8	Very Good
           7	Good
           6	Above Average
           5	Average
           4	Below Average
           3	Fair
           2	Poor
           1	Very Poor

Most likely a human realtor rated this "Overall Qual" column, which means it highly likely takes into account many of the other features. It also means that any future house we intend to predict a price for will need this "Overall Qual" feature, which implies that every new house on the market that will be priced with our ML model will still require a human person!

## Save Final DF

In [ ]:
final_df.to_csv('./data/AMES_Final_DF.csv')

---